Disable client-initiated renegotiation by default

richsalz opened the pull request:
This replaces #15051 which I will close.

This PR is *slightly* bigger but it is much more clean and easy to understand.

mattcaswell commented on include/openssl/ssl.h.in:
```diff
@@ -316,158 +316,135 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
 # define SSL_OP_SAFARI_ECDHE_ECDSA_BUG                   0x00000040U
 # define SSL_OP_IGNORE_UNEXPECTED_EOF                    0x00000080U
-
+    /* See below SSL_OP_ALLOW_CLIENT_RENEGOTIATION           0x00000100U */
```
Why is this one out of order? I think it should be defined here in this list

richsalz replied:
I thought it more useful to keep related options together, but the current `0x000....U` syntax makes that hard. That's why I'd like to use the `1<<xxx` syntax. But that's discussed in #15145 so I moved it.

mattcaswell commented on include/openssl/ssl.h.in:
```diff
@@ -316,158 +316,135 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
 # define SSL_OP_ENABLE_KTLS                              0x00000008U
-
 # define SSL_OP_TLSEXT_PADDING                           0x00000010U
+# define SSL_OP__UNUXED                                  0x00000020U
```
I think this should just be a comment noting that it is unused. Adding a special define means that people might start relying on it being there. If we end up using a `uint64_t` for the OP value then there will be lots of unused slots and it doesn't seem sensible to have a define for each one.

mattcaswell commented on test/helpers/ssltestlib.c:
```diff
@@ -695,7 +695,9 @@ int create_ssl_ctx_pair(OSSL_LIB_CTX *libctx, const SSL_METHOD *sm,
-        else if (!TEST_ptr(serverctx = SSL_CTX_new_ex(libctx, NULL, sm)))
+        else if (!TEST_ptr(serverctx = SSL_CTX_new_ex(libctx, NULL, sm))
+            || !TEST_true(SSL_CTX_set_options(serverctx,
+                    SSL_OP_ALLOW_CLIENT_RENEGOTIATION)))
```
Indentation seems odd here

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -2277,39 +2277,47 @@ int SSL_get_key_update_type(const SSL *s)
-    if ((s->options & SSL_OP_NO_RENEGOTIATION)) {
-        ERR_raise(ERR_LIB_SSL, SSL_R_NO_RENEGOTIATION);
+    if (SSL_is_server(s)
+            && (s->options & SSL_OP_ALLOW_CLIENT_RENEGOTIATION) == 0)
```
This seems wrong. It appears to prevent server initiated renegotiation (on the server side) if the option is not set.

richsalz replied:
Yes, this is in the wrong place. It should be, as you said, around line 1384 of `statem_server.c`

I'll address all your feedback by Monday; today's a holiday/wellness-day for work.

mattcaswell reviewed (changes requested):
I was expecting a change in statem_srvr.c here:

https://github.com/openssl/openssl/blob/22d1138fe2fde9a16e80b81de1d848ae6fa879ef/ssl/statem/statem_srvr.c#L1378-L1393

The changes in the tests seem to be mainly about fixing the tests that were broken by this. There doesn't seem to be a test to check that client initiated reneg doesn't work by default but server initiated reneg does.

mattcaswell commented:
> pushed an updated commit

Did you really? I don't see it...

> Looks like both of those involve the TLSProxy, about which I know nothing. Any clues appreciated while I start reading that code.

My guess is something like this might help:

```` diff
diff --git a/test/recipes/70-test_renegotiation.t b/test/recipes/70-test_renegotiation.t
index 256964266c..1cf14a444e 100644
--- a/test/recipes/70-test_renegotiation.t
+++ b/test/recipes/70-test_renegotiation.t
@@ -36,6 +36,7 @@ my $proxy = TLSProxy::Proxy->new(
 
 #Test 1: A basic renegotiation test
 $proxy->clientflags("-no_tls1_3");
+$proxy->serverflags("-client_renegotiation");
 $proxy->reneg(1);
 $proxy->start() or plan skip_all => "Unable to start up Proxy for tests";
 plan tests => 4;
@@ -45,6 +46,7 @@ ok(TLSProxy::Message->success(), "Basic renegotiation");
 $proxy->clear();
 $proxy->filter(\&reneg_filter);
 $proxy->clientflags("-no_tls1_3");
+$proxy->serverflags("-client_renegotiation");
 $proxy->reneg(1);
 $proxy->start();
 ok(TLSProxy::Message->fail(), "No client SCSV");
@@ -58,7 +60,7 @@ SKIP: {
     $proxy->filter(undef);
     $proxy->ciphers("DEFAULT:\@SECLEVEL=0");
     $proxy->clientflags("-no_tls1_3 -cipher AES128-SHA:\@SECLEVEL=0");
-    $proxy->serverflags("-no_tls1_3 -no_tls1_2");
+    $proxy->serverflags("-no_tls1_3 -no_tls1_2 -client_renegotiation");
     $proxy->reneg(1);
     $proxy->start();
     my $chversion;
@@ -87,6 +89,7 @@ SKIP: {
     $proxy->clear();
     $proxy->filter(\&sigalgs_filter);
     $proxy->clientflags("-tls1_2");
+    $proxy->serverflags("-client_renegotiation");
     $proxy->reneg(1);
     $proxy->start();
     ok(TLSProxy::Message->fail(), "client_sig_algs instead of sig_algs");
diff --git a/test/recipes/70-test_sslmessages.t b/test/recipes/70-test_sslmessages.t
index 3f57af62d5..befc4c7e9e 100644
--- a/test/recipes/70-test_sslmessages.t
+++ b/test/recipes/70-test_sslmessages.t
@@ -239,6 +239,7 @@ checkhandshake($proxy, checkhandshake::CLIENT_AUTH_HANDSHAKE,
 #Test 7: A handshake with a renegotiation
 $proxy->clear();
 $proxy->clientflags("-no_tls1_3");
+$proxy->serverflags("-client_renegotiation");
 $proxy->reneg(1);
 $proxy->start();
 checkhandshake($proxy, checkhandshake::RENEG_HANDSHAKE,
````

richsalz commented:
> Didn't see commit.

Oops.  Forgot to get return  :)

I applied your patch and get `Error with command: "-client_renegotiation"` which i think i can figure out.

richsalz commented:
Thanks very much @mattcaswell!  I still need to write another two tests, right?

richsalz commented on include/openssl/ssl.h.in:
```diff
@@ -316,158 +316,135 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
-/* Use Cisco's "speshul" version of DTLS_BAD_VER
- * (only with deprecated DTLSv1_client_method())  */
+    /*
+     * Use Cisco's "speshul" version of DTLS_BAD_VER
```
Yes, modified.

richsalz commented:
fixup commit pushed.  will work on adding the new tests now.

mattcaswell reviewed (approved):
I would have preferred not to have to add a new command line option just for a test - but ok.

t8m commented on apps/lib/s_cb.c:
```diff
@@ -1233,12 +1233,10 @@ int config_ctx(SSL_CONF_CTX *cctx, STACK_OF(OPENSSL_STRING) *str,
-                           flag, arg);
-            else
-                BIO_printf(bio_err, "Error with command: \"%s\"\n", flag);
+            BIO_printf(bio_err, "Call to SSL_CONF_md(%s, %s) failed\n",
```
Typo: SSL_CONF_cmd

t8m commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -58,9 +58,15 @@ Use server and not client preference order when determining which cipher suite,
 =item B<-legacyrenegotiation>
 
-permits the use of unsafe legacy renegotiation. Equivalent to setting
+Permits the use of unsafe legacy renegotiation. Equivalent to setting
```
The uppercase of the first word is missing for other options.You might correct them all if you are fixing this particular one. But not required of course.

richsalz replied:
There were only two uppercase fixes needed.

t8m replied:
Still need to push changes?

richsalz replied:
i waited for 'make test' to finish :)

t8m commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -70,13 +76,19 @@ B<SSL_OP_NO_RENEGOTIATION>.
 =item B<-no_resumption_on_reneg>
 
-set SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION flag. Only used by servers.
+set B<SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION> flag. Only used by servers.
```
For example here.

t8m commented on ssl/ssl_lib.c:
```diff
@@ -2277,39 +2277,41 @@ int SSL_get_key_update_type(const SSL *s)
+ * Can we accept a renegotiation request?  If yes, set the flag and
+ * return 1 if yes. If not, raise error and return 0.
+ */
+static int can_renegotiation(SSL *s)
```
`can_renegotiate` would sound more English to me. :grin: 

Also, I'd rather keep the duplication of `s->renegotiate = 1;` and keep the function read-only to align with the name.

t8m reviewed (changes requested):
Some nits

richsalz commented:
> I would have preferred not to have to add a new command line option just for a test - but ok.

I could not think of any other way to do it.  Happy to create a followon PR if you have any flash of insight. I do think the flag could be useful for testing clients, but that's rationalization :)  And we do have a options in the list command that are used only for the `cmd-nits` target. :)

richsalz commented:
fixup commit pushed addressing @t8m's feedback. To  avoid extra mail I just marked the issues as resolved.

t8m reviewed (approved):
@mattcaswell still OK?

mattcaswell reviewed (approved):
Still ok

mattcaswell commented:
> I could not think of any other way to do it. Happy to create a followon PR if you have any flash of insight.

I would have done it in sslapitest.c rather than in a TLSProxy test.

richsalz commented:
> I would have done it in sslapitest.c rather than in a TLSProxy test.

okay, I'll think about it.

richsalz commented:
Can someone change the labels to remove the "pending" and show it's approved?  Thanks.

mattcaswell commented:
Oops. Sorry. Done.

mattcaswell commented:
> I would have done it in sslapitest.c rather than in a TLSProxy test.

Actually an even better place might be in:

https://github.com/openssl/openssl/blob/master/test/ssl-tests/17-renegotiate.cnf.in

richsalz commented:
Thanks for the pointer.  I'll look at that, too, but post-alpha-freeze. :)

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

mattcaswell commented:
Ah....there's a non-trivial rebase required due to the 64-bit op PR

richsalz commented:
By design, the ssl.h.in conflict is easily handled since that PR had all of this PR's changes in it :)  Rebase coming up.

richsalz commented:
Rebased, squashed, and pushed. As mentioned above, the conflcit in `ssl.h.in` was handled by just taking the master version, and `CHANGES.md` was trivial. No other changes were made.

mattcaswell commented:
Thanks. @t8m, since this ended up being trivial after all I think we probably don't need to reset the timer for it? Do you agree?

t8m commented:
Sure, merging just now.

t8m commented:
Merged to master. Thank you for the contribution!
